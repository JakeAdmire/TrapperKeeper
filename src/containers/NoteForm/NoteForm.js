import React, { Component } from 'react';
import { NavLink, Redirect } from 'react-router-dom';
import shortid from 'shortid';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { hasError } from '../../actions/index';

import { fetchOptionsCreator } from '../../utility/fetchOptionsCreator'
import { fetchData } from '../../utility/fetchData';
import { fetchAllNotes } from '../../thunks/fetchAllNotes'

import { ListItem } from '../ListItem/ListItem';

export class NoteForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      list: [{
              id: shortid.generate(),
              isComplete: false,
              text: ""
            }],
      toHomePage: false,
    }
  }

  componentDidMount = async () => {
    if(this.props.noteId) {
      await this.findNote(this.props.noteId)
    }
  }

  findNote = async (noteId) => {
    const url = `http://localhost:3001/api/v1/notes/${noteId}`
    try {
      const response = await fetchData(url)
      // console.log('response', response);
      this.setState({
        id: response.id,
        title: response.title,
        list: [...response.list]
      })
    } catch (error) {
      console.log(error.message);
    }
  }

  handleType = (e) => {
    e.preventDefault()
    if(this.props.type === "new-note") {
      this.handlePost()
    } else if(this.props.type === "existing-note") {
      this.handlePut()
    }
  }

  handlePost = async () => {
    const { title, list } = this.state;
    const url = 'http://localhost:3001/api/v1/notes';
    try {
      const options = await fetchOptionsCreator('POST', { title, list })
      await fetchData(url, options)
      this.props.fetchAllNotes('http://localhost:3001/api/v1/notes')
      this.setState({ toHomePage: true })
    } catch (error) {
      this.props.hasError(error.message)
    }
  }

  handlePut = async () => {
    const { title, list } = this.state;
    const url = `http://localhost:3001/api/v1/notes/${this.props.id}`;
    try {
      const options = await fetchOptionsCreator('PUT', { title, list })
      await fetchData(url, options)
      this.setState({ toHomePage: true })
    } catch (error) {
      this.props.hasError(error.message)
    }
  }

  handleTitleChange = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value
    });
  }

  handleItemChange = (e, id) => {
    e.preventDefault();
    const foundItem = this.state.list.find(item => item.id === id);
    foundItem.text = e.target.value;
    this.setState({ list: this.state.list });
    // this.generateNewListItem(e);
  }

  generateNewListItem = (e) => {
    const { value } = e.target;
    if ( value.length === 1 ) console.log('render new item..');
    // if the id doesn't exist, create a new list item
  }

  handleItemDelete = (e, index) => {
    e.preventDefault()
    this.state.list.splice(index, 1)
    this.setState({ list: this.state.list })
  }

  addItem = (e) => {
    e.preventDefault()
    this.setState({ list: [...this.state.list, {
        id: shortid.generate(),
        isComplete: false,
        text: ""
       }
      ]
     })
  }

  deleteNote = async (e) => {
    e.preventDefault();
    const url = `http://localhost:3001/api/v1/notes/${this.props.id}`;
    try {
      const options = await fetchOptionsCreator('DELETE', {})
      await fetchData(url, options)
      this.props.fetchAllNotes('http://localhost:3001/api/v1/notes')
      this.setState({ toHomePage: true })
    } catch (error) {
      this.props.hasError(error.message)
    }
  }

  toggleComplete = (id) => {
    const updatedList = this.state.list.map(item => {
      if(id === item.id) {
        item.isComplete = !item.isComplete
      }
      return item
    });
    this.setState({list: updatedList })
  }

  handleSeperate = () => {
    const { list } = this.state
    let completed = list.filter(item => item.isComplete)
    let uncompleted = list.filter(item => !item.isComplete)
    return {
      completed: {
        items: completed,
      },
      uncompleted: {
        items: uncompleted,
      }
    }
  }

  render() {
    const { toHomePage } = this.state
    if(toHomePage === true){
      return <Redirect to='/' />
    }
    let seperatedList = this.handleSeperate();
    return (
      <div className="Note">
        <section className="Note-Content">
          <NavLink to="/" className="Note-Close">
            <button><i className="fas fa-times"></i></button>
          </NavLink>
          <div className="Note-Form-Container">
            <form className="Note-Form">
              <input type="text"
                     onChange={this.handleTitleChange}
                     placeholder="Title"
                     value={this.state.title}
                     name="title"
                     className="Note-Title"
                     />
              <ul className="ListItems">
                {
                  seperatedList.uncompleted.items.map((item, index) => {
                    return (
                      <ListItem key={item.id} {...item} index={index} toggleComplete={this.toggleComplete} handleItemChange={this.handleItemChange} handleItemDelete={this.handleItemDelete} />
                    )
                  })
                }
              </ul>
              <ul className="ListItems Completed">
                {
                  seperatedList.completed.items.map((item, index) => {
                    return (
                      <ListItem key={item.id} {...item} index={index} toggleComplete={this.toggleComplete} handleItemChange={this.handleItemChange} handleItemDelete={this.handleItemDelete} />
                    )
                  })
                }
              </ul>
              <section className="Note-Options">
                <button className="Note-Save" onClick={this.handleType} type="submit">
                  <i className="fas fa-save"></i>
                </button>
                <button className="Note-Delete" onClick={this.deleteNote}>
                  <i className="fas fa-trash-alt"></i>
                </button>
              </section>
              <button onClick={(e) => this.addItem(e)}>Add List</button>
              <section className="Note-Error"><h2>{this.props.error && this.props.error}</h2></section>
            </form>
          </div>
        </section>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  error: state.error
})

export const mapDispatchToProps = (dispatch) => ({
  hasError: (message) => dispatch(hasError(message)),
  fetchAllNotes: (url) => dispatch(fetchAllNotes(url))
})

NoteForm.propTypes = {
  title: PropTypes.string,
  list: PropTypes.array,
  toHomePage: PropTypes.bool,
  error: PropTypes.string,
  hasError: PropTypes.func
}

export default connect(mapStateToProps, mapDispatchToProps)(NoteForm)
