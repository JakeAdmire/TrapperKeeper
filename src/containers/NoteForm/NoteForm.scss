.Note {
  width: 100%;
  height: 100%;
  top: 0px;
  left: 0px;
  background-color: #e5e5e591;
  position: fixed;
  z-index: 4000;
  transition: opacity 0.218s ease-in 0s;

  &::before {
    content: "";
  }
}

.Note-Content {
  position:fixed;
  background: #ffff;
  width: 100%;
  max-width: 350px;
  border-radius: 5px;
  height: auto;
  padding: 0;
  top: 50%;
  left: 50%;
  transform: translate(-50%,-50%);
  box-shadow: 0px 2px 10px 3px #0000006b;

  .Note-Close {
    position: absolute;
    right: 0;
    top: 0;

    button {
      outline: none;
      cursor: pointer;
      border: none;
      background: #cccccccc;
      color: #fff;
      font-size: 1.25em;
      border-top-right-radius: 5px;
      transition: all .5s ease;

      &:hover {
        color: #a9a9a9;
      }
    }
  }
}

.Note-Form-Container {
  padding: 1em;

  .Note-Form {
    display: flex;
    flex-direction: column;

    .Note-Title {
      font-size: 1.375rem;
      font-weight: 600;
      line-height: 1rem;
      padding-bottom: 5px;
      padding-top: 7px;
      outline: none;
      border: none;
    }

    .ListItems {
      margin: 0;
      padding: 0;

      li {
        list-style-type: none;
        border-top: #d2d2d2 solid 1px;
        border-bottom: #d2d2d2 solid 1px;
        display: flex;
        justify-content: space-between;
        padding: 3px 0;

        label {
          width: 4%;
        }

        input[type=text] {
          outline: none;
          border:none;
          width: 80%;
        }

        button {
          background: none;
          outline: none;
          width: 8%;
          border: none;
          cursor: pointer;
        }

      }


        .container {
          display: block;
          position: relative;
          cursor: pointer;
          font-size: 22px;
          -webkit-user-select: none;
          -moz-user-select: none;
          -ms-user-select: none;
          user-select: none;
        }

        .container input {
          position: absolute;
          opacity: 0;
          cursor: pointer;
          height: 0;
          width: 0;
        }

        .checkmark {
          position: absolute;
          top: 0;
          left: 0;
          height: 16px;
          width: 20px;
          background-color: #eee;
        }

        .container:hover input ~ .checkmark {
          background-color: #ccc;
        }

        .container input:checked ~ .checkmark {
          background-color: #2196F3;
        }

        .checkmark:after {
          content: "";
          position: absolute;
          display: none;
        }

        .container input:checked ~ .checkmark:after {
          display: block;
        }

        .container .checkmark:after {
          left: 8px;
          top: 2px;
          width: 3px;
          height: 7px;
          border: solid white;
          border-width: 0 3px 3px 0;
          -webkit-transform: rotate(45deg);
          -ms-transform: rotate(45deg);
          transform: rotate(45deg);
        }

    }

    .Note-Save {
      margin-top: 2em;
    }

  }
}
