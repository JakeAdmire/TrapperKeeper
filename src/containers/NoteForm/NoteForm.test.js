import React from 'react';
import { shallow } from 'enzyme';
import { NoteForm } from './NoteForm';

const mockShortid = jest.fn().mockReturnValue('12345789');

let mockNote = { id: mockShortid,
  title: 'Worf ToDo',
  list: [
    { id: mockShortid,
      text: 'Eat food',
      isComplete: false
    },
  ]
};

describe('NoteForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<NoteForm type="existing-note" noteId={mockShortid} {...mockNote}/>)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  

})
