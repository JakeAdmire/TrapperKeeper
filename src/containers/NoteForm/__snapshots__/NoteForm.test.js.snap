// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteForm should match the snapshot 1`] = `
<div
  className="Note"
>
  <section
    className="Note-Content"
  >
    <NavLink
      className="Note-Close"
      to="/"
    >
      <button>
        <i
          className="fas fa-times"
        />
      </button>
    </NavLink>
    <div
      className="Note-Form-Container"
    >
      <form
        className="Note-Form"
      >
        <input
          className="Note-Title"
          name="title"
          onChange={[Function]}
          placeholder="Title"
          type="text"
          value=""
        />
        <ul
          className="ListItems"
        >
          <ListItem
            handleItemChange={[Function]}
            handleItemDelete={[Function]}
            id="DiXXZ_LuI"
            index={0}
            isComplete={false}
            key="DiXXZ_LuI"
            text=""
            toggleComplete={[Function]}
          />
        </ul>
        <ul
          className="ListItems Completed"
        />
        <section
          className="Note-Options"
        >
          <button
            className="Note-Save"
            onClick={[Function]}
            type="submit"
          >
            <i
              className="fas fa-save"
            />
          </button>
          <button
            className="Note-Delete"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </section>
        <section
          className="Note-Error"
        >
          <h2 />
        </section>
      </form>
    </div>
  </section>
</div>
`;
